test/core/gprpp/ref_counted_ptr_test.cc

```diff
@@ -250,6 +255,263 @@ TEST(RefCountedPtr, CanPassSubclassToFunctionExpectingSubclass) {
   FunctionTakingSubclass(p);
 }
 
+//
+// WeakRefCountedPtr<> tests
+//
+
+class Bar : public DualRefCounted<Bar> {
+ public:
+  Bar() : value_(0) {}
+
+  explicit Bar(int value) : value_(value) {}
+
+  ~Bar() { GPR_ASSERT(shutting_down_); }
+
+  void Orphan() override { shutting_down_ = true; }
+
+  int value() const { return value_; }
+
+ private:
+  int value_;
+  bool shutting_down_ = false;
+};
+
+TEST(WeakRefCountedPtr, DefaultConstructor) { WeakRefCountedPtr<Bar> bar; }
+
+TEST(WeakRefCountedPtr, ExplicitConstructorEmpty) {
+  WeakRefCountedPtr<Bar> bar(nullptr);
+}
+
+TEST(WeakRefCountedPtr, ExplicitConstructor) {
+  RefCountedPtr<Bar> bar_strong(new Bar());
+  bar_strong->WeakRef().release();
+  WeakRefCountedPtr<Bar> bar(bar_strong.get());
+}
+
+TEST(WeakRefCountedPtr, MoveConstructor) {
+  RefCountedPtr<Bar> bar_strong(new Bar());
+  WeakRefCountedPtr<Bar> bar = bar_strong->WeakRef();
+  WeakRefCountedPtr<Bar> bar2(std::move(bar));
+  EXPECT_EQ(nullptr, bar.get());
```

Comment:
I know that you've defined the move constructor in a specific way, but let me suggest that that's an implementation detail. The semantics of std::move is that the object will be in a valid but unspecified state. So, I don't think it's ok to compare its value to nullptr after a move in general - that becomes a test of the implementation rather than a test of the concept of movability working.

Reply:
I think that for a smart pointer, it's totally reasonable to assume that `std::move()` will reset the value to null.  Even if it's not guaranteed by the spec, there's basically no other implementation that makes sense, because you'd have to maintain a separate data member to indicate validity.

Reply:
Looking over things, I'll agree with you that this is the case with existing smart pointers; both unique_ptr and shared_ptr guarantee that the other will be empty after move assignment..... However, and here is where things get ugly, it turns out that clang-tidy doesn't like this usage pattern. I looked and saw that our internal clang-tidy has marked warnings against the analogous use in RefCountedPtr. Perhaps some sort of NOLINT tag needs to be applied somewhere to protect against this?

Reply:
I'll let this decision be in a follow-up though.

Reply:
I've added NOLINT comments on the appropriate lines.